### ISSUE TITLE
`get_peft_model` and `add_adapter` do not preserve eval mode for newly injected adapter modules

### ISSUE THREAD

**Issue by eSVeeF** [huggingface/peft#3753]
### System Info

- PEFT: `0.21.0` (`main`, commit `514b9e6d`)
- Transformers: `5.3.0`
- Python: `3.12.3`

### Who can help?

@BenjaminBossan 

### Reproduction

I noticed that injecting a LoRA adapter into an already-evaluated model can silently enable adapter dropout.

```python
import torch
from torch import nn
from peft import LoraConfig, get_peft_model


class Tiny(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin = nn.Linear(4, 4, bias=False)

    def forward(self, x):
        return self.lin(x)


config = LoraConfig(
    target_modules=["lin"],
    r=2,
    lora_alpha=2,
    lora_dropout=0.5,
    init_lora_weights=False,
)

model = get_peft_model(Tiny().eval(), config)
layer = model.base_model.model.lin

with torch.no_grad():
    layer.lora_A["default"].weight.fill_(1.0)
    layer.lora_B["default"].weight.fill_(1.0)

x = torch.ones(4, 4)

with torch.no_grad():
    output_1 = model(x)
    output_2 = model(x)

print(model.training)  # True
print(layer.lora_dropout["default"].training)  # True
print(torch.equal(output_1, output_2))  # False
```

The base model was in evaluation mode before adapter injection, so I expected the resulting PEFT model and adapter modules to remain in evaluation mode. Instead, the newly created LoRA dropout module stays in training mode, making repeated inference nondeterministic.

The same happens when adding an adapter to an existing evaluated PEFT model:

```python
model = get_peft_model(
    Tiny(),
    LoraConfig(target_modules=["lin"], lora_dropout=0.0, init_lora_weights=False),
).eval()

model.add_adapter("other", config)
model.set_adapter("other")

print(model.training)  # False
print(model.base_model.model.lin.training)  # False
print(model.base_model.model.lin.lora_dropout["other"].training)  # True
```

In this case, the parent layer is correctly in evaluation mode, but the newly added adapter dropout is not.

### Expected behavior

Expected behavior is that newly injected adapter modules inherit the current training/evaluation state of the model or parent module. This seems different from the existing adapter state-restoration issue in #3507 because the problem occurs during adapter injection rather than `disable_adapter()` context handling.

I’d be happy to open a PR with a shared fix and regression tests for both initial injection and `add_adapter`. Before doing that, would you prefer preserving the full incoming PEFT model’s mode, or explicitly inheriting the training state from each parent module when new adapter modules are created?

**Comment by BenjaminBossan** [comment]
> The base model was in evaluation mode before adapter injection, so I expected the resulting PEFT model and adapter modules to remain in evaluation mode. Instead, the newly created LoRA dropout module stays in training mode, making repeated inference nondeterministic.

This is expected behavior. By default, the PEFT modules are in training mode when calling `get_peft_model`. Think of the most common use case: When loading a pretrained Transformers model, the model will be in eval mode. Users would have to manually put it to train mode if the behavior was different.

> The same happens when adding an adapter to an existing evaluated PEFT model:
> In this case, the parent layer is correctly in evaluation mode, but the newly added adapter dropout is not.

This may or may not be what the user expects, but it is the intended behavior for now: When we call `set_adapter`, we not only make that adapter active, we also make it trainable. It could be argued that this should not be the case, but if we change it, it's reserved for a PEFT v1.0 release, as it would be backward incompatible.

(Note that the training status of the "parent layer", i.e. the `lora.Linear` layer, is ambiguous, as it manages multiple adapters at a time, which can be in different states; we shouldn't make assumptions about what its `.training` should be.)

However, there are some unexpected behaviors around passing `inference_mode` explicitly:

```python
import torch
from torch import nn
from peft import LoraConfig, get_peft_model

class Tiny(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin = nn.Linear(4, 4, bias=False)

    def forward(self, x):
        return self.lin(x)

config = LoraConfig(
    target_modules=["lin"],
    lora_dropout=0.5,
)

model = get_peft_model(
    Tiny(),
    LoraConfig(target_modules=["lin"], lora_dropout=0.0, init_lora_weights=False),
).eval()

# add adapter without inference mode
config = LoraConfig(
    target_modules=["lin"],
    lora_dropout=0.5,
)

model.add_adapter("other", config)
print(model.base_model.model.lin.lora_dropout["other"].training)  # True, expected
model.set_adapter("other", inference_mode=True)
print(model.base_model.model.lin.lora_dropout["other"].training)  # True, should be False

# add adapter with inference mode
config = LoraConfig(
    target_modules=["lin"],
    lora_dropout=0.5,
    inference_mode=True,
)

model.add_adapter("other2", config)
print(model.base_model.model.lin.lora_dropout["other"].training)  # True, should be False
model.set_adapter("other2", inference_mode=False)
print(model.base_model.model.lin.lora_dropout["other2"].training)  # True
model.set_adapter("other2", inference_mode=True)
print(model.base_model.model.lin.lora_dropout["other2"].training)  # True, should be False
```

**Comment by eSVeeF** [comment]
> This is expected behavior. By default, the PEFT modules are in training mode when calling `get_peft_model`.

Got it. Makes sense. Now that you mention it, yeah, I can think of many scenarios that I would find it annoying to have to set it in train mode by hand.

> It could be argued that this should not be the case, but if we change it, it's reserved for a PEFT v1.0 release, as it would be backward incompatible.

Alright.

> However, there are some unexpected behaviors around passing `inference_mode` explicitly

For the explicit `inference_mode=True` case, would you be open to a focused PR just for that? Should I narrow this issue to that case, or would it make more sense to open a separate issue and link the PR there?

Also, should the fix cover all adapter-specific dropout modules across tuners (`boft_dropout`, `randlora_dropout`, etc), or would you prefer to start with LoRA dropout and add focused coverage for `add_adapter(..., inference_mode=True)` and `set_adapter(..., inference_mode=True)`?

**Comment by BenjaminBossan** [comment]
> For the explicit `inference_mode=True` case, would you be open to a focused PR just for that? Should I narrow this issue to that case, or would it make more sense to open a separate issue and link the PR there?

I'd be happy to accept a PR for that. No need for a separate issue.

> Also, should the fix cover all adapter-specific dropout modules across tuners (`boft_dropout`, `randlora_dropout`, etc), or would you prefer to start with LoRA dropout and add focused coverage for `add_adapter(..., inference_mode=True)` and `set_adapter(..., inference_mode=True)`?

I haven't checked what's needed to fix all of those. If it's more or less one location that needs updates, it makes sense to pool those changes. If it's in multiple locations, but we can abstract that to be handled centrally, one PR also makes sense. If it requires lots of changes everywhere, only LoRA is fine.

/peft-triage approved
